core/services/pipeline/common.go

```diff
@@ -427,23 +427,34 @@ func UnmarshalTaskFromMap(taskType TaskType, taskMap interface{}, ID int, dotID
 	return task, nil
 }
 
-func CheckInputs(inputs []Result, minLen, maxLen, maxErrors int) ([]interface{}, error) {
-	if minLen >= 0 && len(inputs) < minLen {
-		return nil, errors.Wrapf(ErrWrongInputCardinality, "min: %v max: %v (got %v)", minLen, maxLen, len(inputs))
+// CheckInputsLen checks that inputs is error free and the expected length.
+func CheckInputsLen(inputs []Result, minLen, maxLen int) error {
+	if minLen > 0 && len(inputs) < minLen {
+		return errors.Wrapf(ErrWrongInputCardinality, "min: %v max: %v (got %v)", minLen, maxLen, len(inputs))
 	} else if maxLen >= 0 && len(inputs) > maxLen {
-		return nil, errors.Wrapf(ErrWrongInputCardinality, "min: %v max: %v (got %v)", minLen, maxLen, len(inputs))
+		return errors.Wrapf(ErrWrongInputCardinality, "min: %v max: %v (got %v)", minLen, maxLen, len(inputs))
 	}
-	var vals []interface{}
-	var errs int
+	return CheckInputs(inputs)
+}
+
+// CheckInputs returns ErrTooManyErrors if inputs contains an Error.
+func CheckInputs(inputs []Result) error {
 	for _, input := range inputs {
 		if input.Error != nil {
-			errs++
-			continue
+			return ErrTooManyErrors
 		}
-		vals = append(vals, input.Value)
 	}
-	if maxErrors >= 0 && errs > maxErrors {
-		return nil, ErrTooManyErrors
+	return nil
+}
+
+// GetInputs returns the Values from inputs and checks that it is error free.
+func GetInputs(inputs []Result) ([]interface{}, error) {
+	vals := make([]interface{}, len(inputs))
+	for i, input := range inputs {
+		if input.Error != nil {
+			return nil, ErrTooManyErrors
```

Comment:
Optional: Probably easier for readability, if you first call CheckInputs(inputs), and only if no errors, then go fetch values without error handling logic here.
It also helps keep the core error checking logic in just 1 place, and avoid duplication.

Reply:
:+1: for readability and de-duplication, but I'm worried about the efficiency of looping twice in a row on the happy path. I'll take another look.

Reply:
Yes, this would loop twice. But on an average, total inputs to a task are in single digits. So looping over ~5 elements twice isn't too bad.
But it is your call, this was just a minor nit.